src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs

```diff
@@ -3696,5 +3656,954 @@ static void Main()
                 );
         }
 
+        [Fact]
+        public void ForWithCircularity1()
+        {
+            string source = @"
+class C
+{
+    static void Main()
+    {
+        for (var (x1, x2) = (1, x1); ; ) { }
+    }
+}
+";
+
+            var comp = CreateCompilationWithMscorlib(source, references: new[] { ValueTupleRef, SystemRuntimeFacadeRef });
+            comp.VerifyDiagnostics(
+                // (6,33): error CS0165: Use of unassigned local variable 'x1'
+                //         for (var (x1, x2) = (1, x1); ; ) { }
+                Diagnostic(ErrorCode.ERR_UseDefViolation, "x1").WithArguments("x1").WithLocation(6, 33)
+                );
+        }
+
+        [Fact]
+        public void ForWithCircularity2()
+        {
+            string source = @"
+class C
+{
+    static void Main()
+    {
+        for (var (x1, x2) = (x2, 2); ; ) { }
+    }
+}
+";
+            var comp = CreateCompilationWithMscorlib(source, references: new[] { ValueTupleRef, SystemRuntimeFacadeRef });
+            comp.VerifyDiagnostics(
+                // (6,30): error CS0165: Use of unassigned local variable 'x2'
+                //         for (var (x1, x2) = (x2, 2); ; ) { }
+                Diagnostic(ErrorCode.ERR_UseDefViolation, "x2").WithArguments("x2").WithLocation(6, 30)
+                );
+        }
+
+        [Fact]
+        public void ForeachIEnumerableDeclarationWithImplicitVarType()
```

Comment:
This is more frequently `ForEach` in the CSharp projects. Consider changing the new instances.